framework/src/materials/ADPiecewiseLinearInterpolationMaterial.C

```diff
@@ -0,0 +1,93 @@
+
+#include "MooseVariableFE.h"
+
+registerADMooseObject("MooseApp", ADPiecewiseLinearInterpolationMaterial);
+
+defineADValidParams(
+    ADPiecewiseLinearInterpolationMaterial,
+    ADMaterial,
+    params.addClassDescription(
+        "Compute a property using a piecewise linear interpolation to define "
+        "its dependence on a variable");
+    params.addRequiredParam<std::string>("property",
+                                         "The name of the property this material will compute");
+    params.addRequiredCoupledVar(
+        "variable",
+        "The name of the variable whose value is used as the abscissa in the interpolation");
+    params.addParam<std::vector<Real>>("x", "The abscissa values");
+    params.addParam<std::vector<Real>>("y", "The ordinate values");
+    params.addParam<std::vector<Real>>("xy_data",
+                                       "All function data, supplied in abscissa, ordinate pairs");
+    params.addParam<Real>("scale_factor",
+                          1.0,
+                          "Scale factor to be applied to the ordinate values"););
+
+template <ComputeStage compute_stage>
+ADPiecewiseLinearInterpolationMaterial<compute_stage>::ADPiecewiseLinearInterpolationMaterial(
+    const InputParameters & parameters)
+  : ADMaterial<compute_stage>(parameters),
+    _prop_name(adGetParam<std::string>("property")),
+    _coupled_var(adCoupledValue("variable")),
+    _scale_factor(adGetParam<Real>("scale_factor")),
+    _property(adDeclareADProperty<Real>(_prop_name))
+{
+  std::vector<Real> x;
+  std::vector<Real> y;
+
+  if ((parameters.isParamValid("x")) || (parameters.isParamValid("y")))
+  {
+    if (!((parameters.isParamValid("x")) && (parameters.isParamValid("y"))))
+      mooseError("In ", _name, ": Both 'x' and 'y' must be specified if either one is specified.");
+
+    if (parameters.isParamValid("xy_data"))
+      mooseError("In ", _name, ": Cannot specify 'x', 'y', and 'xy_data' together.");
+
+    x = adGetParam<std::vector<Real>>("x");
+    y = adGetParam<std::vector<Real>>("y");
+  }
+  else if (parameters.isParamValid("xy_data"))
+  {
+    std::vector<Real> xy = adGetParam<std::vector<Real>>("xy_data");
+    unsigned int xy_size = xy.size();
+    if (xy_size % 2 != 0)
+      mooseError("In ", _name, ": Length of data provided in 'xy_data' must be a multiple of 2.");
+
+    unsigned int x_size = xy_size / 2;
+    x.reserve(x_size);
+    y.reserve(x_size);
+    for (unsigned int i = 0; i < xy_size / 2; ++i)
+    {
+      x.push_back(xy[i * 2]);
+      y.push_back(xy[i * 2 + 1]);
+    }
+  }
+
+  try
+  {
+    _linear_interp = libmesh_make_unique<LinearInterpolation>(x, y);
+  }
+  catch (std::domain_error & e)
+  {
+    mooseError("In ", _name, ": ", e.what());
+  }
+}
+
+template <ComputeStage compute_stage>
+void
+ADPiecewiseLinearInterpolationMaterial<compute_stage>::computeQpProperties()
+{
+  _property[_qp].value() = _scale_factor * _linear_interp->sample(_coupled_var[_qp].value());
```

Comment:
@lindsayad I am trying to implemented `ADPiecewiseLinearInterpolationMaterial` in order to make the tests for elasticity tensors with AD to have perfect jacobians. I am going off of your previous comment where you suggested:
```
_some_prop[_qp].value() = _linear_interpolator.sample(_T[_qp].value());
_some_prop[_qp].derivatives() = _linear_interpolator.sampleDerivative(_T[_qp].value()) * _T[_qp].derivatives();
```
I get the general idea, but I keep getting compile errors:
```
/Users/topher/projects/moose/framework/src/materials/ADPiecewiseLinearInterpolationMaterial.C:89:17: error: member reference base type 'double' is not a structure or union
  _property[_qp].value() = _scale_factor * _linear_interp->sample(_coupled_var[_qp].value());
  ~~~~~~~~~~~~~~^~~~~~
```
If I drop the `_property[_qp].value()` to just `_property[_qp]`, I get a similar compile error,
```
/Users/topher/projects/moose/framework/src/materials/ADPiecewiseLinearInterpolationMaterial.C:89:76: error: member reference base type 'const double' is not a structure or union
  _property[_qp] = _scale_factor * _linear_interp->sample(_coupled_var[_qp].value());
                                                          ~~~~~~~~~~~~~~~~~^~~~~~
```
I don't think this has to be a specified <RESIDUAL>, since you are using `.value()` in `ADCoupledMaterial`...What am I missing here?

Reply:
You will almost certainly have to write a method specialization for this. You could get around the `.value()` calls by using a free function `raw_value` that we have in MetaPhysicL, but we're pretty stuck on the `.derivatives()` call. So `ADCoupledMaterial` looks like this:

```c++
template <ComputeStage compute_stage>
void
ADCoupledMaterial<compute_stage>::computeQpProperties()
{
  _regular_mat_prop[_qp] = 4.0 * _coupled_var[_qp].value();
  _ad_mat_prop[_qp] = 4.0 * _coupled_var[_qp];
}

template <>
void
ADCoupledMaterial<RESIDUAL>::computeQpProperties()
{
  _regular_mat_prop[_qp] = 4.0 * _coupled_var[_qp];
  _ad_mat_prop[_qp] = 4.0 * _coupled_var[_qp];
}
```
The latter specializes for `<RESIDUAL>` and you'll note that there are no `.value()` calls. So you could do the same thing in your new material. The "general" template can represent the `JACOBIAN` calculations and then you can specialize the template for the `<RESIDUAL>` calculations. Or you can have the residual code be in the general template and then specialize the `<JACOBIAN>`; it doesn't make a difference really.

Reply:
Oh okay, makes sense. Would it make more sense to do a <RESIDUAL> and <JACOBIAN> to be clear, or do you have to have a general template.

Reply:
Well, I can get it to compile and run, but the jacobian's are still way off. Here's the updated `ADPiecewiseLinearInterpolationMaterial`
```
template <ComputeStage compute_stage>
void
ADPiecewiseLinearInterpolationMaterial<compute_stage>::computeQpProperties()
{
  _property[_qp] = _scale_factor * _linear_interp->sampleDerivative(_coupled_var[_qp].value()) * _coupled_var[_qp];
}

template <>
void
ADPiecewiseLinearInterpolationMaterial<RESIDUAL>::computeQpProperties()
{
  _property[_qp] = _scale_factor * _linear_interp->sample(_coupled_var[_qp]);
}
```

Which has the general template for the `JACOBIAN`. When using this to test the jacobian in modules/combined/test/tests/thermal_elastic/ad-thermal_elastic.i`, the jacobian blows up
```
||J - Jfd||_F/||J||_F = 2.10052, ||J - Jfd||_F = 6.21256e+06
```

I'll keep digging...

Reply:
Argg!!! Something is jacked up with the residual calculation in `ADPiecewiseLinearInterpolationMaterial`. In `modules/combined/test/tests/thermo_mech/ad-youngs_modulus_function_temp.i`, the coupled temperature `_coupled_var[_qp]` hits a `nan` after between the initial linear and first linear residual. This occurs in an , 
- all AD input
- no AD input besides `ADPiecewiseLinearInterpolationMaterial`
- ADVariableValue/adCoupledValue and VariableValue/coupledValue
- Various BCs
- Various temp kernels
- Various meshes
- Doesn't happen for non-AD `PiecewiseLinearInterpolationMaterial`
- Doesn't happen for fully AD input without AD `PiecewiseLinearInterpolationMaterial`

It's something in `ADPiecewiseLinearInterpolationMaterial`. I know your crammed @lindsayad and @dschwen , but do you guys see something weird in my implementation, or have come across something similar? It's driving me crazy!

Reply:
K I'll take a look

Reply:
That test seems to run ok with the commit I just pushed

Reply:
Yeah that should do it! I'm suprised you need to do the .value() calculation in the jacobian evaluation. I get max of 5e-7 on one test, but that should be good to go. Thanks @lindsayad !!